src/celeritas/optical/detail/PreGenExecutor.hh

```diff
@@ -0,0 +1,100 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/detail/PreGenExecutor.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "celeritas/global/CoreTrackView.hh"
+#include "celeritas/optical/CerenkovPreGenerator.hh"
+#include "celeritas/optical/OpticalGenData.hh"
+#include "celeritas/optical/ScintillationPreGenerator.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+// LAUNCHER
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data.
+ */
+struct PreGenExecutor
+{
+    inline CELER_FUNCTION void
+    operator()(celeritas::CoreTrackView const& track);
+
+    NativeCRef<OpticalPropertyData> const properties;
+    NativeCRef<CerenkovData> const cerenkov;
+    NativeCRef<ScintillationData> const scintillation;
+    NativeRef<OpticalGenStateData> const state;
+    OpticalBufferSize size;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data.
+ */
+CELER_FUNCTION void PreGenExecutor::operator()(CoreTrackView const& track)
+{
+    CELER_EXPECT(state);
+
+    using DistId = ItemId<OpticalDistributionData>;
+
+    auto tsid = track.track_slot_id();
+    auto& cerenkov_dist = state.cerenkov[DistId(size.cerenkov + tsid.get())];
+    auto& scintillation_dist
+        = state.scintillation[DistId(size.scintillation + tsid.get())];
+
+    // clear distribution data
+    cerenkov_dist = {};
+    scintillation_dist = {};
+
+    auto sim = track.make_sim_view();
+    auto optmat_id
+        = track.make_material_view().make_material_view().optical_material_id();
+    if (sim.status() == TrackStatus::inactive || !optmat_id
```

Comment:
I don't think the track status will ever be inactive if the `Executor` is called: the `TrackExecutor` should filter out inactive tracks. If a track *is* inactive, then you might have just accessed invalid memory by getting the optical material ID...

Reply:
Isn't that only true if it's a conditional executor, i.e. constructed with `make_active_track_executor()`? The memory should still be valid even if the track is inactive. I used a regular `TrackExecutor` here because we need to clear any old distribution data in the buffers before compacting with `remove_if`.

Reply:
Oh ok, you're right about the conditional executor. But it still could access uninitialized memory right? If a track has never been initialized, then:
```c++
track.make_material_view().make_material_view().optical_material_id()
```
could obtainan uninitialized/invalid material ID, and then crash when trying to read the optical material from the corresponding material view.

Reply:
Ahh ok I see what you mean: we'd be trying to construct a `MaterialView` with an invalid material ID. You're right, let me fix that...

Reply:
Yeah, I ran into a similar thing once with the cross section calculation.